portal/src/components/roles-and-groups/list/UserDetailsScreenRoleListContainer.tsx

```diff
@@ -0,0 +1,186 @@
+    RolesListQueryDocument,
+    {
+      variables: {
+        pageSize: 0,
+        searchKeyword: "",
+      },
+      fetchPolicy: "network-only",
+    }
+  );
+
+  const [searchKeyword, setSearchKeyword] = useState<string>("");
+  const isSearch = searchKeyword !== "";
+  const [offset, setOffset] = useState(0);
+
+  const onChangeOffset = useCallback((offset) => {
+    setOffset(offset);
+  }, []);
+
+  const onChangeSearchKeyword = useCallback(
+    (e?: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
+      if (e === undefined) {
+        return;
+      }
+      const value = e.currentTarget.value;
+      setSearchKeyword(value);
+      // Reset offset when search keyword was changed.
+      setOffset(0);
+    },
+    []
+  );
+  const onClearSearchKeyword = useCallback(() => {
+    setSearchKeyword("");
+  }, []);
+
+  const [isAddRoleDialogHidden, setIsAddRoleDialogHidden] = useState(true);
+  const showAddRoleDialog = useCallback(
+    () => setIsAddRoleDialogHidden(false),
+    []
+  );
+  const hideAddRoleDialog = useCallback(
+    () => setIsAddRoleDialogHidden(true),
+    []
+  );
+
+  const groupRoles = useMemo(() => {
+    const groupsRolesTable: Record<string, UserRolesListItem> = {};
+    user.groups?.edges?.forEach((edge) => {
+      const group = edge?.node;
+      if (group?.roles?.edges == null) {
+        return;
+      }
+      group.roles.edges.forEach((roleEdge) => {
+        const role = roleEdge?.node;
+        if (role == null) {
+          return;
+        }
+        if (role.key in groupsRolesTable) {
+          groupsRolesTable[role.key].groups.push(group);
+        } else {
+          const roleWithGroups = {
+            ...role,
+            groups: [group],
+          };
+          groupsRolesTable[role.key] = roleWithGroups;
+        }
+      });
+    });
+
+    return Object.entries(groupsRolesTable).map(([_, value]) => value);
+  }, [user.groups?.edges]);
+
+  const userRoles: UserRolesListItem[] = useMemo(() => {
+    return (
+      user.roles?.edges?.flatMap((e) => {
+        if (e?.node) {
+          return [{ ...e.node, groups: [] }];
+        }
+        return [];
+      }) ?? []
+    );
+  }, [user.roles?.edges]);
+
+  const combinedRoles: UserRolesListItem[] = useMemo(() => {
+    return [...groupRoles, ...userRoles];
+  }, [groupRoles, userRoles]);
+
+  const filteredCombinedRoles = useMemo(() => {
+    if (isSearch) {
+      return searchRoles(combinedRoles, searchKeyword);
+    }
+
+    return combinedRoles.slice(offset, offset + pageSize);
+  }, [isSearch, combinedRoles, offset, searchKeyword]);
+
+  if (error != null) {
+    return <ShowError error={error} onRetry={refetch} />;
+  }
+
+  if (loading) {
+    return <ShowLoading />;
+  }
+
+  const totalCount = rolesQueryData?.roles?.totalCount ?? 0;
+
+  if (totalCount === 0) {
+    return <RolesEmptyView />;
+  }
+
+  return (
+    <>
+      <section className={cn("flex flex-col h-full", className)}>
```

Comment:
Prefer not to use `h-full` here, as mentioned in https://github.com/authgear/authgear-server/pull/4013/files#r1531547090